### DISCUSSION TITLE
How do you share envrionment variables?

### DISCUSSION THREAD

**Discussion by baptisteArno** [vercel/turborepo#320] [Help] [answered]
Hi! I have a monorepo with 2 Next.js apps. I'm wondering, what's the best practice for sharing environment variables on those 2 applications?

**Comment by KeKs0r** [comment]
My answer is completely unrelated to turborepo, but we manage environment variabels with doppler, where you can have a config per app. And also share config between apps. So you can basically mimick the structure you have in code.
https://www.doppler.com/

**Reply by baptisteArno** [reply]
Thank you for your answer! 😊
It's probably a bit overkill for my use case as I just need to aggregate env variables into one file for localhost only.

I'm using `dotenv-cli` and adding it to every npm script but that's a bit redundant.

**Reply by KeKs0r** [reply]
Okay, then just for inspiration 2 approaches I have taken in the past:

### Shared Config Package with runtime loading
One shared package for configuration, that you import, which also takes care of loading the `.env` file. So when it is imported the first time (by checking wether the env vars are already set).
```
require('dotenv').config()
```

### Prefixing scripts 
You seem to be doing a version of this. But instead of doing it with every script, you can prefix the program / cli you are running instead of every script.
To prefix all commands that are run by yarn. I for example have used it in the past to execute nx with more memory.
```
    "scripts": {
        "nx": "NODE_OPTIONS='--max_old_space_size=8192' nx",
    }
```
So whenever I run `yarn nx`, it is run with that. You can do the same to prefix your command with dotenv-cli and then e.g. run turbo always with dotenv beforehand.

**Comment by anthonyshew** [comment, marked as answer]
We've updated our documentation for best practices in monorepos when handling environment variables: https://turbo.build/repo/docs/crafting-your-repository/using-environment-variables#best-practices

**Reply by muradsofi** [reply]
@anthonyshew This best practice works well for app-level environment variables. However, it doesn’t apply when setting up a self-hosted Turborepo remote cache.

In this case, the environment variables need to be set for the monorepo itself—not for each individual app. If you're running multiple Turbo commands, managing these variables becomes a poor DX. I also opened a related discussion here: [https://github.com/vercel/turborepo/discussions/10824](https://github.com/vercel/turborepo/discussions/10824?utm_source=chatgpt.com)
explaining why the suggested best practice doesn’t fit this use case.
